cer-graphql/index.js

```diff
@@ -93,26 +101,77 @@ async function createServer() {
                 return forwardReqToContentful(args, context, info);
             } else { // If the user is not signed, do further request checking
 
-                /**
-                 * Get list of requested fields. The location of these differs depending on whether
-                 * this is the resolver for a collection or a single resource.
-                 */
-                let requestedFields = (info.fieldName.endsWith('Collection') ?
-                    info.fieldNodes[0].selectionSet.selections[0].selectionSet.selections :
-                    info.fieldNodes[0].selectionSet.selections)
-                    .map(x => x.name.value);
+                // GraphQL introspection fields, these are used by GraphQL to query metadata
+                const GRAPHQL_INTROSPECTION_FIELDS = [
+                    '__Schema',
+                    '__Type',
+                    '__TypeKind',
+                    '__typename',
+                    '__Field',
+                    '__InputValue',
+                    '__EnumValue',
+                    '__Directive'
+                ];
 
                 // Check whether the user has requested only public fields
                 const ALWAYS_PUBLIC_FIELDS = [
                     'title',
                     'summary',
                     'name',
-                    'ssoProtected'
+                    'ssoProtected',
+                    'searchable',
+                    'linkedFrom',
+                    'slug',
+                    'icon',
+                    ...GRAPHQL_INTROSPECTION_FIELDS
                 ];
 
-                let userOnlyQueryingPublicFields = requestedFields
-                    .every(y => ALWAYS_PUBLIC_FIELDS.includes(y));
+                var requestedFields = []; // List of fields requested, populated in the recursive function below
+
+                let recursive_iteration_count = 0; // Tracks the number of recursive iterations
+                const MAX_RECURSIVE_ITERATIONS = 500; // The maximum allowed number of recursive iterations
+                var getRequestedFields = function (obj) {
```

Comment:
Awesome work on this recursive function!

Reply:
Perhaps it could be pulled out to the top level instead of inside the createServer function.

Reply:
Thanks man! Only took me... 2.5 days 😭 

I originally had it as a separate function, but moved it back in to the `createServer` function because I couldn't get it to play nicely with our integration tests:

https://github.com/UoA-eResearch/hub-stack/blob/e8d37d93aa3828db566f501e49a69fe03ef83e1e/cer-graphql/index.js#L273-L275

We export that stuff here so that the `createServer` function is available in our integration tests file:

https://github.com/UoA-eResearch/hub-stack/blob/e8d37d93aa3828db566f501e49a69fe03ef83e1e/cer-graphql/tests/integration-tests.test.js#L7-L14